hermes/probe/read/read.go

```diff
@@ -0,0 +1,130 @@
+// Before it creates and stores a file it logs an intent to do so in the target storage system's journal.
+// It verifies that the creation and storage process was successful.
+// Finally, it updates the filenames map in the target's journal and record the exit status in the logger.
+// Arguments:
+//          ctx: it carries deadlines and cancellation signals that might orinate from the main probe located in probe.go.
+//          target: contains information about target storage system, carries an intent log in the form of a StateJournal and it used to export metrics.
+//          fileID: the unique identifer of every file, it cannot be repeated. It needs to be in the range [minFileID, maxFileID]. FileID 0 is reserved for a special file called the NIL file.
+//          client: is a storage client. It is used as an interface to interact with the target storage system.
+//          logger: a cloudprober logger used to record the exit status of the ReadFile operation in a target bucket. The logger passed MUST be a valid logger.
+// Returns:
+//          error: an error string with detailed information about the status and fileID. Nil is returned when the operation is successful.
+func ReadFile(ctx context.Context, target *probe.Target, fileID int32, fileSize int, client stiface.Client, logger *logger.Logger) error {
+	if fileID < minFileID || fileID > maxFileID {
+		return fmt.Errorf("invalid argument: fileID = %d; want %d <= fileID <= %d", fileID, minFileID, maxFileID)
+	}
+	bucketName := target.Target.GetBucketName()
+
+	// Verify that the file is present in the State Journal
+	fileName, ok := target.Journal.Filenames[fileID]
+	if !ok {
+		return fmt.Errorf("file with the ID: %d is missing from the State Journal", fileID)
+	}
+	verifyFileExists(ctx, client, bucketName, target, fileName, fileID)
+	start := time.Now()
+	reader, err := client.Bucket(bucketName).Object(fileName).NewReader(ctx)
+	if err != nil {
+		var status metrics.ExitStatus
+		switch err {
+		case storage.ErrObjectNotExist:
+			status = metrics.FileMissing
+		case storage.ErrBucketNotExist:
+			status = metrics.BucketMissing
+		default:
+			status = metrics.ProbeFailed
+		}
+		target.LatencyMetrics.APICallLatency[metrics.APIGetFile][status].Metric(hermesAPILatencySeconds).AddFloat64(time.Now().Sub(start).Seconds())
+		return fmt.Errorf(".%q: could not read file %q: %w", status, fileName, err)
+	}
+	defer reader.Close()
+	h := sha1.New()
+	if _, err := io.Copy(h, reader); err != nil {
+		return fmt.Errorf("checksum calculation failed io.Copy: %w", err)
+	}
+	gotChecksum := fmt.Sprintf("%x", h.Sum(nil))
+	fileNamePrefix := fmt.Sprintf(FileNameFormat, fileID, "")
+	wantChecksum := fileName[len(fileNamePrefix):]
+	if gotChecksum != wantChecksum {
+		return fmt.Errorf("the calculated checksum: %q does not match the checksum in the file name: %q", gotChecksum, wantChecksum)
+	}
+	logger.Infof("verified consistency for object %q in bucket %q", fileName, bucketName)
```

Comment:
Aren't we meant to re-create the file contents using the random byte generator again and check this checksum against the file name and contents read checksum?

Reply:
Added an issue for this:
https://github.com/googleinterns/step224-2020/issues/80

Reply:
That could be an extra feature added in the future!